aten/src/ATen/core/jit_type.h

```diff
@@ -324,25 +324,50 @@ inline c10::optional<T> merge_primitive(
   return c10::optional<T>{};
 }
 
-// `VaryingShape` tracks if individual dimensions or a rank vary across
-// profiled runs. A *varying* or *dynamic* dimension is expressed as
-// an empty c10::optional in `sizes_`. If a rank is dynamic, the entire
-// `sizes_` becomes the empty optional.
+struct CAFFE2_API ShapeSymbol {
+    ShapeSymbol(int64_t val, bool statik = true): value_(val), statik_(statik) {}
+
+    explicit operator int() const { return value_; }
+    ShapeSymbol(const ShapeSymbol& ss) {
+      statik_ = ss.statik_;
+      value_ = ss.value_;
+    }
+    ShapeSymbol(): value_(-1), statik_(false) {}
+    static ShapeSymbol getInvalidSymbol() {return ShapeSymbol(); }
```

Comment:
Seems odd that these can be invalid. It makes it hard to know statically when they are not used.

Reply:
Alternatively, because in every place we accept a shapesymbol , it might be unknown. We can explictly reserve one value of the integer to mean "unknown shape", which is not equal to any other unknown shape. This would prevent having to wrap all ShapeSymbols in optionals. In other words Shape symbols would represent something like:

* `?` - completely unknown, encoding 0
* `3` - literally size 3, encoding 3
* `a` - the same size as other `a`, encoding -1